src/java/main/org/apache/zookeeper/common/ZKTrustManager.java

```diff
@@ -0,0 +1,137 @@
+        this.x509ExtendedTrustManager = x509ExtendedTrustManager;
+        this.hostnameVerificationEnabled = hostnameVerificationEnabled;
+        this.shouldVerifyClientHostname = shouldVerifyClientHostname;
+
+        hostnameVerifier = new DefaultHostnameVerifier();
+    }
+
+    @Override
+    public X509Certificate[] getAcceptedIssuers() {
+        return x509ExtendedTrustManager.getAcceptedIssuers();
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket) throws CertificateException {
+        if (hostnameVerificationEnabled && shouldVerifyClientHostname) {
+            performHostVerification(socket.getInetAddress(), chain[0]);
+        }
+        x509ExtendedTrustManager.checkClientTrusted(chain, authType, socket);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket) throws CertificateException {
+        if (hostnameVerificationEnabled) {
+            performHostVerification(socket.getInetAddress(), chain[0]);
+        }
+        x509ExtendedTrustManager.checkServerTrusted(chain, authType, socket);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine engine) throws CertificateException {
+        if (hostnameVerificationEnabled && shouldVerifyClientHostname) {
+            try {
+                performHostVerification(InetAddress.getByName(engine.getPeerHost()), chain[0]);
+            } catch (UnknownHostException e) {
+                throw new CertificateException("failed to verify host", e);
+            }
+        }
+        x509ExtendedTrustManager.checkServerTrusted(chain, authType, engine);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngine engine) throws CertificateException {
+        if (hostnameVerificationEnabled) {
+            try {
+                performHostVerification(InetAddress.getByName(engine.getPeerHost()), chain[0]);
+            } catch (UnknownHostException e) {
+                throw new CertificateException("failed to verify host", e);
+            }
+        }
+        x509ExtendedTrustManager.checkServerTrusted(chain, authType, engine);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        x509ExtendedTrustManager.checkClientTrusted(chain, authType);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        x509ExtendedTrustManager.checkServerTrusted(chain, authType);
+    }
+
+    private void performHostVerification(InetAddress inetAddress, X509Certificate certificate) throws CertificateException {
+        try {
+            hostnameVerifier.verify(inetAddress.getHostAddress(), certificate);
```

Comment:
So we have a few hostname checks that we could consider. Please let me know if I am missing any.

1. Check that the address/hostname of the machine matches what is listed on the certificate
2. Check that the address/hostname of the machine matches what is expected from zoo.cfg (this may also require reading the sid from the packet sent when a connection is established). 
3. Check that the address/hostname listed on the certificate matches what is expected from zoo.cfg

This patch only performs check 1. I think that check 3 is something @geek101 mentioned earlier and we agreed this can be moved to another patch as it is more complicated. 

I think it is also possible to implement check 2. This can exist entirely outside of tls/ssl support and is something we should consider based on its own merits. 

I think having checks 1 and 2 would be plenty secure and we would not need check 3.

Let me know what you think @hanm.